mlir/lib/Interfaces/ControlFlowInterfaces.cpp

```diff
@@ -177,7 +177,10 @@ LogicalResult detail::verifyTypesAlongControlFlowEdges(Operation *op) {
 
     SmallVector<RegionBranchTerminatorOpInterface> regionReturnOps;
     for (Block &block : region)
-      if (auto terminator = dyn_cast<RegionBranchTerminatorOpInterface>(
+      if (block.mightHaveTerminator() &&
+              dyn_cast<RegionBranchTerminatorOpInterface>(
+                  block.getTerminator());
+          auto terminator = dyn_cast<RegionBranchTerminatorOpInterface>(
```

Comment:
This kind of `if` statement is somewhat controversial (see https://discourse.llvm.org/t/rfc-code-style-of-initializer-in-if-statement/65357) and the way you've written it causes interface lookup to be performed twice.

I think it'd be neater to just write it as:
```cpp
if (block.mightHaveTerminator())
  if(auto terminator = dyn_cast<RegionBranchTerminatorOpInterface>(block.getTerminator())
    ...
```